### PR TITLE
1.10/Ability to download workers sorted by home centre name and then worker name

### PR DESCRIPTION
**Pull request by AlinChindea**
### Trello card 👉 [https://trello.com/c/hs9uwODu/1563-ability-to-download-worker-list-with-a-click-of-a-button-ee-vv](https://trello.com/c/hs9uwODu/1563-ability-to-download-worker-list-with-a-click-of-a-button-ee-vv)

❗ Only deals with one item on that card. Another PR dealt with the bulk of the logic of this card. See #400 

`sortBy()` multiple columns is an issue in Laravel - [https://github.com/laravel/ideas/issues/11](https://github.com/laravel/ideas/issues/11) so I took to googling and followed this suggestion from a [StackOverflow answer](https://stackoverflow.com/questions/25451019/what-is-the-syntax-for-sorting-an-eloquent-collection-by-multiple-columns/25451441#25451441)

### CODE REVIEW

File: app/Http/Controllers/Service/Admin/CentreUsersController.php [hunk 1]
```diff
@@ -223,7 +223,10 @@ private function syncCentres($request, CentreUser $cu)
 
     public function download()
     {
-        $workers = CentreUser::get()->sortBy('homeCentre.name');
+        $workers = CentreUser::get()->sortBy(function($centreUser) {
+            return sprintf('%-12s%s', $centreUser->homeCentre->name, $centreUser->name);
```
Comment (charlesstrange2) [inline app/Http/Controllers/Service/Admin/CentreUsersController.php:227]: ah! hashing the  two together and sorting them
I read that as 
parameter 1, left jsutified, first 12 characters,
Parameter 2, not justified. 

why the twelve characters, and why the left justify?

Comment (charlesstrange2) [reply app/Http/Controllers/Service/Admin/CentreUsersController.php:227]: Also,
could you return 
"string1#string2"
using the delimeter "#"?

Comment (AlinChindea) [reply app/Http/Controllers/Service/Admin/CentreUsersController.php:227]: I don't really get it what's going on here. But here goes an explanation I gleaned from teh interwebs. If you need the sortBy against multiple columns, you probably need to space pad them to ensure that "ABC" and "DEF" comes after "AB" and "DEF", hence the sprint right padded for each column up to the column's length (at least for all but the last column)

Comment (AlinChindea) [reply app/Http/Controllers/Service/Admin/CentreUsersController.php:227]: > Also,
> could you return
> "string1#string2"
> using the delimeter "#"?

this doesn't seem to do anything. my version, below 👇 

`$workers = CentreUser::get()->sortBy(function() {
            return 'homeCentre.name#name';
        });`

Comment (AlinChindea) [reply app/Http/Controllers/Service/Admin/CentreUsersController.php:227]: alas, found an easier solution. had to read an article I skipped the first time around...

Comment (AlinChindea) [reply app/Http/Controllers/Service/Admin/CentreUsersController.php]: it's counterintuitive but if you have users collection and you want to sort by last name, first name, you'd think you can do `dd($users->sortBy('last_name')->sortBy('first_name'));`

actually, you need to put them in reverse order
`dd($users->sortBy('first_name')->sortBy('last_name'));`

👀 🕵️‍♂️

Comment (charlesstrange2) [reply app/Http/Controllers/Service/Admin/CentreUsersController.php]: something like
```
$workers = CentreUser::get()->sortBy(function ($centreUser) {
  return $centreUser->homeCentre->name . "#" . $centreUser-name
}
```

File: app/Http/Controllers/Service/Admin/CentreUsersController.php [hunk 2]
```diff
@@ -223,9 +223,7 @@ private function syncCentres($request, CentreUser $cu)
 
     public function download()
     {
-        $workers = CentreUser::get()->sortBy(function($centreUser) {
-            return sprintf('%-12s%s', $centreUser->homeCentre->name, $centreUser->name);
-        });
+        $workers = CentreUser::get()->sortBy('name')->sortBy('homeCentre.name');
```
Comment (charlesstrange2) [inline app/Http/Controllers/Service/Admin/CentreUsersController.php:226]: as per https://github.com/laravel/ideas/issues/11#issuecomment-329866327 ?
give it a go.

Comment (AlinChindea) [reply app/Http/Controllers/Service/Admin/CentreUsersController.php:226]: yes, it's similar to that but then there's another comment after [https://github.com/laravel/ideas/issues/11#issuecomment-329970458](https://github.com/laravel/ideas/issues/11#issuecomment-329970458) that says that the reverse sorting works accidentally and is based on undefined behaviour. I saw it it here, too [https://medium.com/@codebyjeff/the-keys-are-the-key-in-laravel-multiple-field-sortby-60cd5bf3f78b](https://medium.com/@codebyjeff/the-keys-are-the-key-in-laravel-multiple-field-sortby-60cd5bf3f78b)

so I don't know which one to choose... thoughts?

**Pull request merged**